### PR TITLE
{start,stop}-mycroft.sh: make POSIX sh compatible and fix shellcheck issues

### PR DESCRIPTION
**Pull request by PureTryOut**
## Description
Makes the start and stop scripts POSIX sh compatible, and fixes any `shellcheck` issues found.

This doesn't really depend on #2802, but without it `bin/mycroft-start` and `bin/mycroft-stop` will break. Thus, #2802 should be merged first.

## How to test
Run all the usual variants of this test like normal (e.g. `./start-mycroft.sh`) and see if anything behaves differently than before

## Contributor license agreement signed?
CLA [x] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

**Comment by PureTryOut** [comment]
Hmm, not sure why Travis fails.

```
$ ./start-mycroft.sh unittest
./start-mycroft.sh: 85: .: Can't open /home/travis/build/MycroftAI/mycroft-core/.venv/bin/activate
```

This works fine for me locally.

**Comment by forslund** [comment]
The python env in Travis exists already so the .venv/bin/activate is not actually used here. Earlier the command failed and carried on with the already active venv. (See here for example https://travis-ci.org/github/forslund/mycroft-core/jobs/635713266#L1410)...so either add `|| true` or perhaps the better way is to do a check if a venv already is active...

The following oneliner should be best according to stack overflow
`python -c 'import sys;sys.exit(not (hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)))'`

Rather long but still a single line...

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

**Comment by PureTryOut** [comment]
I do not care if it's one line or more to be honest, but then again I'm not sure how to split that up into multiple :joy: 

Thanks for the explanation, I'll add that check.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/sh
```
Comment (fermulator) [inline start-mycroft.sh:1]: we should still be using `env`
```
#!/usr/bin/env sh
```

File: start-mycroft.sh [hunk 2]
```diff
@@ -14,15 +14,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-SOURCE="${BASH_SOURCE[0]}"
+# This script is never sourced but always directly executed, so this is safe to do
```
Comment (fermulator) [inline start-mycroft.sh:17]: I mean ... technically we should embed all "work" in a `main()` function to avoid sourcing collisions ... but I agree not a huge issue.

File: start-mycroft.sh [hunk 3]
```diff
@@ -14,15 +14,16 @@
...
+SOURCE="$0"
 
 script=${0}
 script=${script##*/}
-cd -P "$( dirname "$SOURCE" )"
+cd -P "$( dirname "$SOURCE" )" || exit
```
Comment (fermulator) [inline start-mycroft.sh:22]: silent failure? ideally log the error here

File: start-mycroft.sh [hunk 4]
```diff
@@ -77,51 +78,54 @@ function name-to-script-path() {
     esac
 }
 
-function source-venv() {
-    # Enter Python virtual environment, unless under Docker
```
Comment (fermulator) [inline start-mycroft.sh:81]: what is the requirement/reasoning for not using python venv within docker? (is it pre-loaded?)

File: start-mycroft.sh [hunk 5]
```diff
@@ -77,51 +78,54 @@ function name-to-script-path() {
...
-    if [ ! -f "/.dockerenv" ] ; then
-        source ${VIRTUALENV_ROOT}/bin/activate
+source_venv() {
+    # Enter Python virtual environment, unless under Docker or when a virtualenv is already active
+    virtualenv=$(python3 -c 'import sys; sys.stdout.write("1") if (hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)) else sys.stdout.write("0")')
```
Comment (fermulator) [inline start-mycroft.sh:83]: i tried this on my `hass` venv environment 

OUTSIDE venv:
```
fermulator@hass:~$ python3 -c 'import sys; sys.stdout.write("1") if (hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)) else sys.stdout.write("0")'
0 
```

INSIDE venv
```
python3 -c 'import sys; sys.stdout.write("1") if (hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)) else sys.stdout.write("0")'
1
```

:heavy_check_mark:

File: start-mycroft.sh [hunk 6]
```diff
@@ -131,39 +135,39 @@ function launch-background() {
...
-    if [[ "${1}" == "bus" ]] ; then
+    if [ "${1}" = "bus" ] ; then
         echo "CAUTION: The Mycroft bus is an open websocket with no built-in security"
         echo "         measures.  You are responsible for protecting the local port"
         echo "         8181 with a firewall as appropriate."
     fi
 
     # Launch process in background, sending logs to standard location
-    python3 -m ${_module} $_params >> /var/log/mycroft/${1}.log 2>&1 &
+    python3 -m ${_module} "$_params" >> /var/log/mycroft/"${1}".log 2>&1 &
```
Comment (fermulator) [inline start-mycroft.sh:145]: i would enclose the ENTER blob log file path in quotes (cleaner)

```
... >> "/var/log/mycroft/${1}.log" ...
```

File: start-mycroft.sh [hunk 7]
```diff
@@ -131,39 +135,39 @@ function launch-background() {
...
         auto_update="false"
     fi
-    if [ "$auto_update" == "true" ] ; then
+    if [ "$auto_update" = "true" ] ; then
         # Check github repo for updates (e.g. a new release)
         git pull
     fi
 
-    if [ ! -f .installed ] || ! md5sum -c &> /dev/null < .installed ; then
+    if [ ! -f .installed ] || ! md5sum -c 2>&1 > /dev/null < .installed ; then
```
Comment (fermulator) [inline start-mycroft.sh:168]: what's the reason to also shove `stderr` to `/dev/null`?

File: start-mycroft.sh [hunk 8]
```diff
@@ -180,80 +184,80 @@ function check-dependencies() {
...
         pytest "$@"
         ;;
     "skillstest")
-        source-venv
+        source_venv
         pytest test/integrationtests/skills/discover_tests.py "$@"
         ;;
     "vktest")
-        source "$DIR/bin/mycroft-skill-testrunner" vktest "$@"
+        bash "$DIR/bin/mycroft-skill-testrunner" vktest "$@"
```
Comment (fermulator) [inline start-mycroft.sh:245]: shouldn't this script define its own shebang? (ideally we wouldn't force `bash` from here)

File: stop-mycroft.sh [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/sh
```
Comment (fermulator) [inline stop-mycroft.sh:1]: similar comments on this file with `start-mycroft.sh` (fix throughout) as applicable

File: stop-mycroft.sh [hunk 10]
```diff
@@ -40,37 +41,37 @@ function help() {
...
             fi
         done
 
-        if process-running $1 ; then
+        if process_running "$1" ; then
             echo "failed to stop."
             pid=$( pgrep -o -f "python3 (.*)-m mycroft.*${1}" )            
-            echo -n "  Killing $1 (${pid})..."
-            kill -9 ${pid}
+            printf "  Killing %s (%s)..." "$1" "${pid}"
```
Comment (fermulator) [inline stop-mycroft.sh:73]:
```
printf "  Killing %s (%s)..." "${1}" "${pid}"
```

also, when I tried this I don't know why it trails a `$`

```
$ FOO=blah
$ BAR=23

$ printf "  Killing %s (%s)..." "$FOO" "$BAR"
  Killing blah (23)...$ ```

**Review by fermulator — CHANGES_REQUESTED** [review]

File: start-mycroft.sh (hunk 4 above)
Comment (forslund) [reply start-mycroft.sh:81]: If I recall correctly they wanted to skip this step in the docker since it isn't strictly needed (the docker environment is enough of a sandbox). However I believe the current docker setup (in [docker-mycroft](https://github.com/mycroftai/docker-mycroft) is currently using the venv

File: start-mycroft.sh (hunk 8 above)
Comment (forslund) [reply start-mycroft.sh:245]: It does have it's own shebang so bash could be omitted... @krisgesling do you remember if there was any thought behind sourcing it? (Making it the same level as the rest of the function calls?)

Comment (krisgesling) [reply start-mycroft.sh:245]: May have been related to something else I was trying and just got left in there. I can't think of any reason that it would be needed. Remove it I think.

File: stop-mycroft.sh (hunk 10 above)
Comment (forslund) [reply stop-mycroft.sh:73]: Is that your prompt maybe?

There should be a `\n` at the end of the string

File: start-mycroft.sh (hunk 1 above)
Comment (PureTryOut) [reply start-mycroft.sh:1]: I don't see a reason why. For one, env is a GNU Coreutils thing, and I'd like this to work on systems without those installed. Secondly, we can assume that on POSIX systems `/bin/sh` is always available and it's a POSIX compatible shell. If it's not, that's broken on the distro in question and not something that we should care about, as that's just a system not being POSIX.

File: start-mycroft.sh (hunk 2 above)
Comment (PureTryOut) [reply start-mycroft.sh:17]: Agreed, I might do that in the future but that's out of the scope of this PR for now.

File: start-mycroft.sh (hunk 6 above)
Comment (PureTryOut) [reply start-mycroft.sh:145]: That's a personal preference but one I can get behind, resolved.

File: stop-mycroft.sh (hunk 10 above)
Comment (PureTryOut) [reply stop-mycroft.sh:73]: Strange, Bash doesn't seem to append the `\n` but zsh does. I'll add a `\n` anyway to force it.

File: start-mycroft.sh (hunk 7 above)
Comment (PureTryOut) [reply start-mycroft.sh:168]: No reason really, that's just a habit of mine I guess. Good call, I'll change it.

File: start-mycroft.sh (hunk 3 above)
Comment (PureTryOut) [reply start-mycroft.sh:22]: That's... A bit hard to do. It'll exit with whatever the return code was of `cd` which should be enough indication in itself on what went wrong, but to actually log something is a bit tricky.

https://github.com/koalaman/shellcheck/wiki/SC2242

I'm not entirely sure how to capture the result of `cd` and check it later to then log it in some conditional statement.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh (hunk 3 above)
Comment (forslund) [reply start-mycroft.sh:22]: I do think cd will log the failure (atleast in my tests it does). And as you said exit-status is preserved so this should be fine

File: stop-mycroft.sh [hunk 11]
```diff
@@ -90,30 +91,28 @@ OPT=$1
 shift
```
Comment (krisgesling) [inline stop-mycroft.sh:91]: It seems that `sh` doesn't handle shifting when there are no arguments.
```
./mycroft-core/stop-mycroft.sh: 91: shift: can't shift that many
```

So perhaps we need a conditional on here?

**Review by krisgesling — COMMENTED** [review]
Hey, this looks great, thanks!

Only one small issue with `shift` that I think we need to resolve.

Thanks @fermulator and @forslund for reviewing and testing this already too!

File: stop-mycroft.sh (hunk 11 above)
Comment (PureTryOut) [reply stop-mycroft.sh:91]: That is strange, that shouldn't be the case. Afaik if none argument is specified, it should just do `N+1`.
Shellcheck doesn't complain either. How did you get this error? I can't reproduce it personally.

I could do `shift 1` but that seems redundant. What conditional would you suggest?

Comment (krisgesling) [reply stop-mycroft.sh:91]: Apparently the expected behaviour in [ash](https://linux.die.net/man/1/ash):
> Shift the positional parameters n times. A shift sets the value of $1 to the value of $2, the value of $2 to the value of $3, and so on, decreasing the value of $# by one. If n is greater than the number of positional parameters, shift will issue an error message, and exit with return status 2.

What about:
```
OPT=$1
if [ $# -gt 0 ] ; then
    shift
fi
```

Comment (PureTryOut) [reply stop-mycroft.sh]: Yup thanks for that pointer. My system doesn't run Ash (`/bin/sh` points to Busybox) for me so I didn't catch this.

Modified it, could you try again?

File: stop-mycroft.sh [hunk 12]
```diff
@@ -87,33 +88,33 @@ result=0  # default, no change
 
 
 OPT=$1
-shift
+if [ $# -gt 0]; then
```
Comment (krisgesling) [inline stop-mycroft.sh:91]: little typo, missing the space between `0 ]`

Comment (PureTryOut) [reply stop-mycroft.sh:91]: :see_no_evil:

File: start-mycroft.sh [hunk 13]
```diff
@@ -180,80 +184,80 @@ function check-dependencies() {
...
-if [[ "${1}" == "restart" ]] || [[ "${_opt}" == "restart" ]] ; then
+if [ "${1}" = "restart" ] || [ "${_opt}" = "restart" ] ; then
     _force_restart=true
-    if [[ "${_opt}" == "restart" ]] ; then
+    if [ "${_opt}" = "restart" ] ; then
         # Support "start-mycroft.sh restart all" as well as "start-mycroft.sh all restart"
         _opt=$1
     fi
     shift
 fi
```
Comment (krisgesling) [inline start-mycroft.sh:194]: Sorry I just realised we also run into the same problem here. It's not intended behaviour to use this script without args, but I'd rather we show the help info if someone does try to use it incorrectly.

I guess the easiest method would be to throw a conditional on both? But it's getting late here so I may just be getting lazy...

Comment (PureTryOut) [reply start-mycroft.sh:200]: The key bit being "It's not intended behaviour to use this script without args".

I checked if `$#` equals 0 now, and if it does show the help and exit. If it's at least 1 or more, both shifts should be successful now I think?

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh (hunk 13 above)
Comment (krisgesling) [reply start-mycroft.sh:200]: The other edge case would be `mycroft-start restart`
which now fails with a shift error when called using start-mycroft.sh
and directs to update dependencies for bin/mycroft-start

Comment (PureTryOut) [reply start-mycroft.sh]: I'm not sure how that still happens but ok, I added a check for shift within the restart block too.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh [hunk 14]
```diff
@@ -179,81 +183,90 @@ function check-dependencies() {
...
+if [ "${1}" = "restart" ] || [ "${_opt}" = "restart" ] ; then
     _force_restart=true
-    if [[ "${_opt}" == "restart" ]] ; then
+    if [ "${_opt}" = "restart" ] ; then
         # Support "start-mycroft.sh restart all" as well as "start-mycroft.sh all restart"
         _opt=$1
     fi
-    shift
+
+    if [ $# -gt 1 ]; then
```
Comment (krisgesling) [inline start-mycroft.sh:200]: I think we want this as `-gt 0` as the first param has already been shifted off.

File: test.sh [hunk 15]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+echo "$#"
```
Comment (krisgesling) [inline test.sh:3]: Looks like this got committed by accident

File: start-mycroft.sh [hunk 16]
```diff
@@ -14,15 +14,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-SOURCE="${BASH_SOURCE[0]}"
+# This script is never sourced but always directly executed, so this is safe to do
+SOURCE="$0"
```
Comment (krisgesling) [inline start-mycroft.sh:18]: Don't know how I glossed over this previously but the script is sourced in the `bin/mycroft-start` helper command.

This means the $PWD will get set to `mycroft-core/bin` which causes issues in `check_dependencies` when it is reading from files stored in the parent `mycroft-core` directory.

**Review by krisgesling — COMMENTED** [review]
Sorry, looks like we aren't quite done...

File: start-mycroft.sh (hunk 16 above)
Comment (PureTryOut) [reply start-mycroft.sh:18]: Hmm yes you're right. POSIX doesn't have an easy way to do this, this will get ugly...

Comment (PureTryOut) [reply start-mycroft.sh:18]: It seems it's just not possible to determine the path of a source script from within itself using just POSIX features...
https://stackoverflow.com/questions/29832037/how-to-get-script-directory-in-posix-sh/29835459#29835459

So, not sure how to resolve this. `pwd` seems to always return the directory in which the script is located (at least for me with Busybox, GNU Coreutils seem to do the same), no matter if it's sourced or not, but I can not find anything about that behaviour online so I'm not sure if we should rely on it. Otherwise it would be as easy as `SOURCE=$(pwd)`.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh (hunk 16 above)
Comment (krisgesling) [reply start-mycroft.sh:17]: Thinking out loud here... in the helper commands could we set `$BASH_SOURCE=$(pwd)` or an equivalent variable. Then test for it's existence in the start/stop scripts rather than assume it is there or not?

Comment (PureTryOut) [reply start-mycroft.sh:17]: What helper commands do you mean? `bin/mycroft-start`?

File: stop-mycroft.sh [hunk 17]
```diff
@@ -40,37 +41,37 @@ function help() {
...
-function end-process() {
-    if process-running $1 ; then
+end_process() {
+    if process_running "$1" ; then
         # Find the process by name, only returning the oldest if it has children
         pid=$( pgrep -o -f "python3 (.*)-m mycroft.*${1}" )
-        echo -n "Stopping $1 (${pid})..."
-        kill -SIGINT ${pid}
+        printf "Stopping %s (%s)..." "$1" "${pid}"
+        kill -SIGINT "${pid}"
```
Comment (forslund) [inline stop-mycroft.sh:57]: Was looking at this to see if it could be merged got an issue with this:

```
Stopping skills (6297)..../stop-mycroft.sh: 57: kill: Illegal option -S
failed to stop.
  Killing skills (6297)...
killed.
Stopping audio (6300)..../stop-mycroft.sh: 57: kill: Illegal option -S
failed to stop.
  Killing audio (6300)...
killed.
Stopping speech (6303)..../stop-mycroft.sh: 57: kill: Illegal option -S
failed to stop.
  Killing speech (6303)...
killed.
Stopping enclosure (6306)..../stop-mycroft.sh: 57: kill: Illegal option -S
failed to stop.
  Killing enclosure (6306)...
killed.
Stopping messagebus.service (17057)..../stop-mycroft.sh: 57: kill: Illegal option -S
failed to stop.
  Killing messagebus.service (17057)...
killed.
```

seems like kill in sh doesn't work with signal names?

Comment (PureTryOut) [reply stop-mycroft.sh]: Is `-SIGINT` equivelant to `-s SIGINT`? I believe sh just requires the `-s` option before specifying what signal to send. I just pushed a change, could you try again?

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: stop-mycroft.sh (hunk 17 above)
Comment (forslund) [reply stop-mycroft.sh]: Now it's `Stopping skills (5419)..../stop-mycroft.sh: 57: kill: invalid signal number or name: SIGINT` `kill -l` in sh suggests it should be just `INT`

Comment (PureTryOut) [reply stop-mycroft.sh]: How strange, fixed.

Comment (forslund) [reply stop-mycroft.sh]: Yes, sh seems to be Dash on my machine.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: stop-mycroft.sh (hunk 17 above)
Comment (forslund) [reply stop-mycroft.sh]: Working now. :+1:

File: start-mycroft.sh (hunk 16 above)
Comment (forslund) [reply start-mycroft.sh:17]: Looked a bit on this and the helper commands in bin that uses start-mycroft.sh and stop-mycroft.sh seem to fail (when called from outside of the mycroft-core root folder). I wonder if this was what Kris meant. but I'm not sure.

I think this is due to the fact that the files are sourced and not executed as a script. for me it seems to work if instead of sourcing the start-mycroft.sh the script is executed.

For example in mycroft-start the change would be
```diff
diff --git a/bin/mycroft-start b/bin/mycroft-start
index 65723c2c37..bda0967048 100755
--- a/bin/mycroft-start
+++ b/bin/mycroft-start
@@ -18,4 +18,4 @@ SOURCE="${BASH_SOURCE[0]}"
 cd -P "$( dirname "$SOURCE" )"/..
 DIR="$( pwd )"
 
-. "$DIR/start-mycroft.sh" $@
+"$DIR/start-mycroft.sh" $@
```

Comment (PureTryOut) [reply start-mycroft.sh:17]: I've just gone ahead and made this depend on #2802, as that'll completely replace `bin/mycroft-start` for a new script anyway. Maybe at some point we can get rid of these location dependent scripts entirely with just `bin/mycroft-{start,stop}`, who knows.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2686)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2686/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2686/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2686/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2686/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2686/logs/enclosure.log)

**Comment by PureTryOut** [comment]
Just rebased this. @krisgesling anything holding this back?

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2686))

File: start-mycroft.sh (hunk 1 above)
Comment (fermulator) [reply start-mycroft.sh:1]: sure thing! sounds good (if a dev system doesn't have coreutils installed that's odd; but agreed that `/bin/sh` is most portable)

**Review by fermulator — APPROVED** [review]

File: start-mycroft.sh [hunk 18]
```diff
@@ -179,83 +179,91 @@ function check-dependencies() {
...
-fi
 
+    if [ $# -gt 0 ]; then
+	    shift
+    fi
+fi
 
-if [[ ! "${_opt}" == "cli" ]] ; then
+if [ ! "${_opt}" = "cli" ] ; then
     check-dependencies
```
Comment (forslund) [inline start-mycroft.sh]: I think this should be replaced with the `check_dependencies`

**Review by forslund — DISMISSED** [review]
Looks really good and seems to work well. I did find a place where the `check-dependencies`call hadn't been updated (L202)

File: start-mycroft.sh [hunk 19]
```diff
@@ -141,29 +141,29 @@ function launch-background() {
...
         auto_update="false"
     fi
-    if [ "$auto_update" == "true" ] ; then
+    if [ "$auto_update" = "true" ] ; then
         # Check github repo for updates (e.g. a new release)
         git pull
     fi
 
-    if [ ! -f .installed ] || ! md5sum -c &> /dev/null < .installed ; then
+    if [ ! -f .installed ] || ! md5sum -c 2>&1 > /dev/null < .installed ; then
```
Comment (forslund) [inline start-mycroft.sh]: For me to silence this completely when md5sum detects a difference is detected I had to change the order of the redirects:
```sh
md5sum -c > /dev/null 2>&1 < .installed
```

**Review by krisgesling — APPROVED** [review]
With Ake's final catches addressed, this seems good to go!

:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:
